Add shared agent guidance (AGENTS.md, CLAUDE.md, AI.md) (#7655)
elastic/docs-content: AGENTS.md

Pages are [MyST Markdown](https://elastic.github.io/docs-builder/syntax/) with Elastic extensions, built by [docs-builder](https://elastic.github.io/docs-builder/). After a merge to `main`, the published site refreshes within minutes, followed by a cache refresh.

## What belongs here, and what doesn't

This repo hosts the cross-product, narrative user documentation for the products and versions already described. Some content lives elsewhere:

- **Reference docs that sit next to the code they describe.** Reference content generated from or maintained alongside a codebase stays in that code repo. API references are the clearest case: they're generated from OpenAPI specs, most often in `elastic/kibana` or `elastic/elasticsearch-specification`, not authored as Markdown here. The same holds for other in-repo reference material that ships with the product.

> - **Legacy-version content.** Documentation for versions earlier than the ones this repo covers, essentially Elastic v8 and before, lives in the respective code repositories, on older branches, in AsciiDoc, and is not built by docs-builder. Update those repos directly when relevant, but only for a clear product behavior change or critical fix.

Reviewer: arguably this should be in the main repo `README.md`

Author: yeah good point, I never read the readme, how silly, I moved some stuff over there

## Core principles

> - **Verify before you draft.** For anything the product actually does, such as UI labels, defaults, and behavior, the source of truth is the product's code repo at `HEAD`, not the issue or PR description. Never publish something the model inferred without checking source. Because these are cumulative docs, also confirm a statement holds for earlier relevant product minors versions when there is any doubt: A page must stay true for users on any version that this repo documents.

Reviewer 2: ```suggestion
- **Verify before you draft.** For anything the product actually does, such as UI labels, defaults, and behavior, the source of truth is the product's code repo at `HEAD`, not the issue or PR description. Never publish something the model inferred without checking source. Because these are cumulative docs, also confirm a statement holds for earlier relevant product minor versions when there is any doubt: A page must stay true for users on any version that this repo documents.
```

- **Find the canonical home for content.** Search for a page that already covers the topic before adding a new one.
- **Place content where it belongs, once.** Put each detail in its single most correct page, not everywhere the feature is mentioned. When a feature appears in several places, keep them consistent and update all of them. Reuse shared content with a snippet, though a single duplication across a couple of pages beats maintaining an include. See [content types](contribute-docs/content-types/index.md).
- **Cumulative docs.** These docs serve every supported version at once. Preserve existing content and scope new content with `applies_to` rather than overwriting. See the [cumulative-docs guide](contribute-docs/how-to/cumulative-docs/index.md).
- **Don't assume existing content already follows every rule.** Consistency with surrounding pages is a good default, but the existing docs are not guaranteed correct. When content obviously diverges from the documented best practices, prefer fixing or improving it over matching the divergence.
- **Write for the reader, not the PR.** Translate developer jargon into user language. One precise sentence beats three vague ones.

> ## Repo layout

Reviewer: again, could be in main README?

Author: yeah good point, I never read the readme, how silly, I moved some stuff over there

| Path | What it documents |
|---|---|
| `get-started/` | Onboarding and first-run content across products. Fundamental knowledge about Elastic |
| `solutions/` | Docs for solutions and serverless project types packaged on top of Elasticsearch and Kibana: Search, Observability, Security, Vector DB |
| `explore-analyze/` | Home for core Kibana features such as Discover, dashboards, visualizations, alerting, AI features, ML |

> | `deploy-manage/` | Deploying, scaling, securing, and operating clusters, deployments, and projects |

Reviewer 2: ```suggestion
| `deploy-manage/` | Deploying, scaling, securing, and operating orchestrators, clusters, deployments, and projects |
```

| `manage-data/` | Ingesting, transforming, and lifecycle-managing data |
| `reference/` | Reference material such as settings, APIs, CLIs, and config |
| `troubleshoot/` | Troubleshooting and diagnostics |
| `release-notes/` | Release notes and changelogs for Elastic Stack and Serverless. Produced from code changes, not part of regular docs edits. |
| `cloud-account/` | Elastic Cloud account and billing |
| `extend/` | Extending and integrating with Elastic |
| `contribute-docs/` | The public contribution guide: style, content types, cumulative docs, syntax. The canonical source this file points to. |

> | `serverless/`, `archive.md` | Serverless landing content and archived material |

Author: ```suggestion
| `get-started/` | Onboarding and first-run content across products. Fundamental knowledge about Elastic |
| `solutions/` | Docs for solutions and serverless project types packaged on top of Elasticsearch and Kibana: Search, Observability, Security, Vector DB |
| `explore-analyze/` | Home for core Kibana features such as Discover, dashboards, visualizations, alerting, AI features, ML |
| `deploy-manage/` | Deploying, scaling, securing, and operating clusters, deployments, and projects |
| `manage-data/` | Ingesting, transforming, and lifecycle-managing data |
| `reference/` | Reference material such as settings, APIs, CLIs, and config that isn't naturally sitting in other code repos |
| `troubleshoot/` | Troubleshooting and diagnostics |
| `release-notes/` | Release notes and changelogs for Elastic products. Generally generated from code changes. |
| `cloud-account/` | Elastic Cloud account and billing |
| `extend/` | Extending and integrating with Elastic, for external developers |
| `contribute-docs/` | The public contribution guide: style, content types, cumulative docs, syntax. The canonical source this file points to. |
| `serverless/`, `archive.md` | Serverless landing content and archived material. This content should not be updated. It's legacy and no longer published. |
```

- Second person, present tense, active voice. Sentence case for headings.
- Bold UI labels and controls, such as **Save**, **Add panel**, and **Settings**.
- Separate navigation steps with ` → `, as in **Add** → **Controls** → **Variable control**.
- Use MyST directives per the [syntax quick reference](contribute-docs/syntax-quick-reference.md). Keep content in prose where it reads naturally, and don't stack admonitions.
- Write links with the [docs-builder link syntax](https://elastic.github.io/docs-builder/syntax/links/), including the cross-repo `<repo>://` form. Validate them with the `crosslink-validator` skill.

## When you open a PR

Opening a PR builds a preview automatically. The link appears on the PR and updates within minutes. Before requesting review, check your changes for:

- **Content type compliance**: the right structure and depth for the page's type.
- **applies_to and cumulative-docs compliance**: correct scoping for versions and deployments.
- **Style compliance**: apply the [Vale](contribute-docs/vale-linter.md) recommendations where relevant.
- **Accuracy against the product code**: labels, defaults, and behavior verified at `HEAD`.

> Preview locally with `docs-builder serve`, or run `docs-builder` with no argument for a full build that surfaces errors ([build locally](contribute-docs/locally.md)). Run Vale on changed files. When you move, rename, or delete a page that has already been published, add the redirect in `redirects.yml`. For AI-assisted contributions specifically, read [`AI.md`](AI.md).

Reviewer 3: I'm wondering how this will work in practice: are we saying that agents should automatically run `docs-builder serve` or `docs-builder` whenever they draft content? What if a contributor already has a local build running?

Author: by doing this in their sandbox they can detect new warnings/errors before considering the change ready for review, so it is an automatic way to ensure it's not going to create a PR that fails CI.
If there's already a local build running, it'll find that the port is occupied and either restart it, or do it on another port

Author: In practice this is likely the type of things you'd see in your agent chat

<img width="896" height="62" alt="image" src="https://github.com/user-attachments/assets/bd55028a-3a9d-438c-a29c-834d8d5dcff5" />

Reviewer: ++ to have agent building locally, it's great hygiene 💯

Reviewer: if have local build already running the agent can inspect the rendered docs in browser too 👁️

Reviewer 4: If it has Playwright MCP, which is a stretch. But we might add a line about the option.

## Tooling

The **Elastic docs skills** (from <https://github.com/elastic/elastic-docs-skills>) cover the common tasks. Install them, then reach for the one that fits. When a skill doesn't cover your case, fall back to the contribution guide, as described at the top of this file.

| Task | Skill |
|---|---|

> | Issue to drafted PR, end to end | `accept-docs-quest` |

Reviewer 3: Is this a skill for drafting content? I can't find it in https://github.com/elastic/elastic-docs-skills/tree/main/skills

Author: ```suggestion
```
it snuck it from my local environment, removing for now

- **Legacy-version content.** Documentation for versions earlier than the ones this repo covers, essentially Elastic v8 and before, lives in the respective code repositories, on older branches, in AsciiDoc, and is not built by docs-builder. Update it when a change is relevant to those versions, but not in this repo. Opening pull requests against those targets is encouraged, though the general rule is to do so only for a clear product behavior change or a critical fix.

## Finding a page from its URL

To locate the source file behind a published page, strip `https://www.elastic.co/docs/` from the URL and append `.md`:

```
https://www.elastic.co/docs/explore-analyze/discover/document-explorer
→ explore-analyze/discover/document-explorer.md
```

This is a rule of thumb for finding files, not for writing links. Links in the docs follow their own rules, defined in [the docs-builder link syntax](https://elastic.github.io/docs-builder/syntax/links/).

## Core principles

> - **Verify before you draft.** For anything the product actually does, such as UI labels, defaults, and behavior, the source of truth is the product's code repo at `HEAD`, not the issue or PR description. Never publish something the model inferred without checking source. Because these are cumulative docs, also confirm a statement holds for earlier v9 minors when there is any doubt. A page must stay true for users on any supported v9 minor.

Reviewer 2: not just v9 - min version+ minors. not sure how to simply state this, using [this](https://www.elastic.co/docs/get-started/versioning-availability#find-docs-for-your-product-version) as a source of truth

```suggestion
- **Verify before you draft.** For anything the product actually does, such as UI labels, defaults, and behavior, the source of truth is the product's code repo at `HEAD`, not the issue or PR description. Never publish something the model inferred without checking source. Because these are cumulative docs, also confirm a statement holds for earlier product minor versions when there is any doubt. A page must stay true for users on {{version.<product>.base}} and higher.
```

Author: yeah I tink the hint is sufficient for the agent to understand. Not sure it'd understand {{version.stack.base}} though this is not set locally in this repo

```suggestion
- **Verify before you draft.** For anything the product actually does, such as UI labels, defaults, and behavior, the source of truth is the product's code repo at `HEAD`, not the issue or PR description. Never publish something the model inferred without checking source. Because these are cumulative docs, also confirm a statement holds for earlier relevant product minors versions when there is any doubt: A page must stay true for users on any version that this repo documents.
```

High-signal conventions:

- Second person, present tense, active voice. Sentence case for headings.
- Bold UI labels and controls, such as **Save**, **Add panel**, and **Settings**.
- Separate navigation steps with ` → `, as in **Add** → **Controls** → **Variable control**.
- Use MyST directives per the [syntax quick reference](contribute-docs/syntax-quick-reference.md). Keep content in prose where it reads naturally, and don't stack admonitions.

## Before you open a PR

Opening a PR builds a preview automatically. The link appears on the PR and updates within minutes. Before requesting review, check your changes for:

- **Content type compliance**: the right structure and depth for the page's type.
- **applies_to and cumulative-docs compliance**: correct scoping for versions and deployments.
- **Style compliance**: apply the [Vale](contribute-docs/vale-linter.md) recommendations where relevant.
- **Accuracy against the product code**: labels, defaults, and behavior verified at `HEAD`.

> Preview locally with `docs-builder serve`, or run `docs-builder` with no argument for a full build that surfaces errors ([build locally](contribute-docs/locally.md)). Run Vale on changed files. When you move, rename, or delete a page, add the redirect in `redirects.yml`. For AI-assisted contributions specifically, read [`AI.md`](AI.md).

Reviewer 2: ```suggestion
Preview locally with `docs-builder serve`, or run `docs-builder` with no argument for a full build that surfaces errors ([build locally](contribute-docs/locally.md)). Run Vale on changed files. When you move, rename, or delete a page that has already been published, add the redirect in `redirects.yml`. For AI-assisted contributions specifically, read [`AI.md`](AI.md).
```

Keep this file lean: it points to the [`contribute-docs/`](contribute-docs/index.md) guide rather than restating it. Two resources back it up. When the **Elastic docs skills** are installed, reach for the relevant skill first, since it's the fast path. When a skill doesn't fit or you're unsure, the contribution guide has the deeper context, and unlike the skills it's always present in this repo. If this file and the contribution guide ever disagree, the guide wins, so fix this file.

## About this repo

This is the source for the narrative user documentation published at <https://www.elastic.co/docs/>, built by [docs-builder](https://elastic.github.io/docs-builder/) from MyST Markdown. For the products and versions it covers, what belongs here versus in the code repos, and the repo layout, see the [README](README.md).

## Core principles

> - **Verify before you draft.** For anything the product actually does, such as UI labels, defaults, and behavior, the source of truth is the product's code repo at `HEAD`, not the issue or PR description. Never publish something the model inferred without checking source. Because these are cumulative docs, also confirm a statement holds for earlier relevant product minor versions when there is any doubt: A page must stay true for users on any version that this repo documents.

Reviewer 5: It'd be nice if the agent could flag items that it couldn't verify by reading the issue and corresponding PRs. I prefer to flag unknowns by adding comments in the md file, but I'm open to other methods if folks have different preferences.

Author: hmm yeah i'll make that explicit
```suggestion
- **Verify before you draft.** For anything the product actually does, such as UI labels, defaults, and behavior, the source of truth is the product's code repo at `HEAD`, not the issue or PR description. Never publish something the model inferred without checking source. Because these are cumulative docs, also confirm a statement holds for earlier relevant product minor versions when there is any doubt: A page must stay true for users on any version that this repo documents. Surface important facts that couldn't be verified.
```

- **Find the canonical home for content.** Search for a page that already covers the topic before adding a new one.
- **Place content where it belongs, once.** Put each detail in its single most correct page, not everywhere the feature is mentioned. When a feature appears in several places, keep them consistent and update all of them. Reuse shared content with a snippet, though a single duplication across a couple of pages beats maintaining an include. See [content types](contribute-docs/content-types/index.md).
- **Cumulative docs.** These docs serve every supported version at once. Preserve existing content and scope new content with `applies_to` rather than overwriting. See the [cumulative-docs guide](contribute-docs/how-to/cumulative-docs/index.md).
- **Don't assume existing content already follows every rule.** Consistency with surrounding pages is a good default, but the existing docs are not guaranteed correct. When content obviously diverges from the documented best practices, prefer fixing or improving it over matching the divergence.

> - **Write for the reader, not the PR.** Translate developer jargon into user language. One precise sentence beats three vague ones.

Reviewer 5: How will the agent know the reader/audience that it's writing for? Is it inferring this from the surrounding text or related pages? Or is it doing this in some other way?

Author: That's a content type or skill question, not for the agents file
